### PR TITLE
Make setitem accept 0d np-arrays

### PR DESCRIPTION
**Pull request by guilhermeleobas**
This is an attempt to fix issue #4458. Still a working in progress and some things in my to-do list to explore.

I've updated `SetItemBuffer` to correctly annotate the signature of the operation `ary[idx] = val`, when `val` is an 0-dim numpy array. As of today, Numba is failing to generate LLVM IR code for this operation.

### CODE REVIEW

**Comment by guilhermeleobas** [comment]
I think this PR is ready for a round of reviews

**Comment by esc** [comment]
@guilhermeleobas thanks for pinging us on this, I have updated the label accordingly.

File: numba/core/typing/arraydecl.py [hunk 1]
```diff
@@ -170,7 +170,7 @@ def generic(self, args, kws):
 class SetItemBuffer(AbstractTemplate):
     def generic(self, args, kws):
         assert not kws
-        ary, idx, val = args
+        ary, idx, val = args # ary[idx] = val
```
Comment (stuartarchibald) [inline numba/core/typing/arraydecl.py]: Remove commented code?!

File: numba/np/arrayobj.py [hunk 2]
```diff
@@ -4957,6 +4957,20 @@ def array_to_array(context, builder, fromty, toty, val):
     return val
 
 
+@lower_cast(types.Array, types.Float)
+@lower_cast(types.Array, types.Integer)
+def array0d_to_scalar(context, builder, fromty, toty, val):
+    # a is an array(T, 0d, C)
+    # See: https://github.com/numba/numba/pull/4516
```
Comment (stuartarchibald) [inline numba/np/arrayobj.py]: This information is recorded in git history, it can be removed.

File: numba/np/arrayobj.py [hunk 3]
```diff
@@ -4957,6 +4957,20 @@ def array_to_array(context, builder, fromty, toty, val):
     return val
 
 
+@lower_cast(types.Array, types.Float)
+@lower_cast(types.Array, types.Integer)
+def array0d_to_scalar(context, builder, fromty, toty, val):
+    # a is an array(T, 0d, C)
```
Comment (stuartarchibald) [inline numba/np/arrayobj.py]: This comment should go into `impl`? There's no `a` in this scope?

File: numba/tests/test_unicode_array.py [hunk 4]
```diff
@@ -303,6 +303,25 @@ def test_setitem(self):
...
+        exp2 = pyfunc(x1.copy(), 2, v1)
+        got2 = cfunc(x1.copy(), 2, v1)
+        self.assertPreciseEqual(exp2, got2)
+
+        # --------------
+        x2 = np.zeros(3, dtype=np.int64)
+        v2 = np.array(3, dtype=np.int64)
+        exp3 = pyfunc(x2.copy(), 0, v2)
+        got3 = cfunc(x2.copy(), 0, v2)
+        self.assertPreciseEqual(exp3, got3)
```
Comment (stuartarchibald) [inline numba/tests/test_unicode_array.py]: Why is this here, this is testing unicode arrays?! Perhaps it ought to go in `numba.tests.test_fancy_indexing` along with the other `setitem` tests?

**Comment by stuartarchibald** [comment]
Here's a few tests to consider:
```python
from numba import njit
import numpy as np

def test(test_name, idx, x, y):

    print(test_name.center(80, '-'))
    def foo(idx, a, b):
        a[idx] = b

    pyraised = False
    try:
        tmp1 = np.copy(x)
        foo(idx, tmp1, y)
    except Exception as e:
        print("Python raises", e)
        pyraised = True
    else:
        print("Python result", tmp1)

    nbraised = False
    try:
        tmp2 = np.copy(x)
        njit(foo)(idx, tmp2, y)
    except Exception as e:
        print("Numba raises", e)
        nbraised = True
    else:
        print("Numba result", tmp2)

    if pyraised ^ nbraised:
        print("FAIL")
    else:
        try:
            if not np.allclose(tmp1, tmp2):
                print("FAIL")
            else:
                print("PASS")
        except Exception as e:
            print(e)
            print("FAIL")


test('control test', 0, np.zeros(3), np.array(1.)) # control test, is fine
test('int coerce', 0, np.zeros(3), np.array(1)) # should work, Numba doesn't coerce int
test('chr x3', 0, np.zeros((3,), dtype='<U3'), np.array('a')) # Numba cannot type
test('3d 1 element', 0, np.zeros((3,),), np.array([[[1]]])) # This should work, needs to be done at runtime?
```
it's entirely ok to defer some of the implementation to another PR/later work, especially the last item where I think it's going to require runtime implementation/checking. Thanks for working on this.

File: numba/tests/test_unicode_array.py (hunk 4 above)
Comment (guilhermeleobas) [reply numba/tests/test_unicode_array.py]: Moved to a different file

**Comment by guilhermeleobas** [comment]
Only the first test was successful. I'll try to address tests 2-3 in this PR and 4 in another 4.

**Comment by guilhermeleobas** [comment]
@stuartarchibald, can you give me some help? I think some work must be done for UnicodeCharSeq. In the third case, Numba infers the type of [the access](https://github.com/guilhermeleobas/numba/blob/7780775063ea7ff1f4a9cf8f31a772aaa0a33f30/numba/core/typing/arraydecl.py#L183) `a[idx]` as `UnicodeCharSeq(3)`, whereas I would expect this to be just a single char.

```python
def foo(idx, a, b):
    a[idx] = b
    # type(a) => UnicodeCharSeq(3)
    # type(b) => UnicodeCharSeq(1)
    # type(idx) => int64
```

Numba is inferring the access `a[idx]` as `UnicodeCharSeq(3)`. In the first case for instance, `a` is an `array(float64, 0d, C)` and the access `a[idx]` would be of type `float64`. Does that make sense?

By the way, case two is fixed.

**Comment by guilhermeleobas** [comment]
@stuartarchibald, PR is ready for a review. One test is failing but I don't think this is because of any changes I did.

**Comment by stuartarchibald** [comment]
@guilhermeleobas thanks, test failure is unrelated to this PR https://github.com/numba/numba/pull/5345 should fix.

**Comment by guilhermeleobas** [comment]
Ok, I've merged master with this branch.

**Comment by guilhermeleobas** [comment]
Pinging @hameerabbasi

File: numba/np/arrayobj.py [hunk 5]
```diff
@@ -5047,6 +5047,28 @@ def array_to_array(context, builder, fromty, toty, val):
     return val
 
 
+@lower_cast(types.Array, types.Float)
+@lower_cast(types.Array, types.Integer)
```
Comment (hameerabbasi) [inline numba/np/arrayobj.py:5051]: What about complex floating point numbers?

Comment (guilhermeleobas) [reply numba/np/arrayobj.py:5051]: Done! Thanks for the review

**Comment by guilhermeleobas** [comment]
Hi @stuartarchibald, I think this PR is ready to be merged.

**Comment by guilhermeleobas** [comment]
Hi @stuartarchibald, can you take a look at this PR when you have some free time?

**Comment by ziofil** [comment]
@stuartarchibald, it would be nice to have this PR merged 🥺

File: numba/core/typing/arraydecl.py [hunk 6]
```diff
@@ -215,6 +215,10 @@ def generic(self, args, kws):
                 else:
                     return
             res = val
+        elif isinstance(val, types.Array) and val.ndim == 0 \
+            and self.context.can_convert(val.dtype, res):
+            # val is an array(T, 0d, C), where T is the type of res
```
Comment (stuartarchibald) [inline numba/core/typing/arraydecl.py:220]:
```suggestion
            # val is an array(T, 0d, O), where T is the type of res, O is order
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the fix ups from the initial review, there's a few minor things to resolve and then it should be good for merge. Thanks again.

**Comment by guilhermeleobas** [comment]
Hi @stuartarchibald, I've added tests for implicit casting. Testing `complex(array)` requires a [some changes](https://github.com/numba/numba/blob/3e84cae251f318e6eaed97eb53c56ce67cadbe16/numba/core/typing/builtins.py#L963-L970) to its typing/lowering behavior and I can do this in a future PR.

File: numba/np/arrayobj.py [hunk 7]
```diff
@@ -5047,6 +5047,29 @@ def array_to_array(context, builder, fromty, toty, val):
     return val
 
 
+@lower_cast(types.Array, types.Float)
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/np/arrayobj.py:5147]: Hi @stuartarchibald, I've added tests for implicit casting. Testing `complex(array)` requires a [some changes](https://github.com/numba/numba/blob/3e84cae251f318e6eaed97eb53c56ce67cadbe16/numba/core/typing/builtins.py#L963-L970) to its typing/lowering behavior and I can do this in a future PR.

**Comment by guilhermeleobas** [comment]
Hi @stuartarchibald, CI is green. Can you take a look at this PR when you have some free time

**Comment by ziofil** [comment]
@stuartarchibald could you take a look and merge if ready? I keep using a workaround for turning 0d arrays into scalars in all of my code, but it would be cleaner without it..

**Comment by guilhermeleobas** [comment]
My patch no longer works with Numba master. I'll spend some time in the upcoming weeks trying to fix it.

**Comment by stuartarchibald** [comment]
@guilhermeleobas https://github.com/stuartarchibald/numba/commit/77195ee277693b40857285475000444bc33d9008 should be in the right direction to get CI passing again, feel free to pull in the patch if it's useful.

**Comment by guilhermeleobas** [comment]
CI is all green. Can you take a look @stuartarchibald

File: numba/np/arrayobj.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/np/arrayobj.py:5177]: I think it would be good to get this working to get feature parity across the numerical types, thanks!

[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/np/arrayobj.py:5177]: It turns out that `complex` casting was working. I've also include timedelta64 and datetime64 casts

File: numba/tests/test_fancy_indexing.py [hunk 8]
```diff
@@ -176,6 +176,52 @@ def test_setitem_array(self):
...
+    def test_setitem_0d(self):
+        # Test setitem with a 0d-array
+        pyfunc = setitem_usecase
+        cfunc = jit(nopython=True)(pyfunc)
+
+        x1 = np.zeros(3)
+        v1 = np.array(3.14)
+        exp1 = pyfunc(x1.copy(), 0, v1)
+        got1 = cfunc(x1.copy(), 0, v1)
+        self.assertPreciseEqual(exp1, got1)
```
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:188]: `setitem_usecase` returns `None`, is this just asserting `None == None` ?
https://github.com/numba/numba/blob/07bad806dfaee4f764ed1d7407885da1105cf790/numba/tests/test_fancy_indexing.py#L15-L16

File: numba/tests/test_fancy_indexing.py [hunk 9]
```diff
@@ -176,6 +176,52 @@ def test_setitem_array(self):
...
+        got5 = cfunc(x_ucs.copy(), 0, v_ucs)
+        self.assertPreciseEqual(exp5, got5)
+
+        # --------------
+        # Complex
+        x_cp = np.zeros(3, dtype=np.complex)
+        v_cp = np.array(2+3j, dtype=np.complex)
+        exp5 = pyfunc(x_cp.copy(), 0, v_cp)
+        got5 = cfunc(x_cp.copy(), 0, v_cp)
+        self.assertPreciseEqual(exp5, got5)
```
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:223]: I wonder if the 0-d arrays can be the induction variables of a loop and all the variations can be tested with the same code block calling `setitem_usecase`?

File: numba/np/arrayobj.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/np/arrayobj.py:5177]: Great, thanks.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the update. I've tested this manually against a variety of cases and have found a few things that might need another look. Thanks again for working on this.

File: numba/tests/test_fancy_indexing.py (hunk 8 above)
Comment (guilhermeleobas) [reply numba/tests/test_fancy_indexing.py:188]: You're right! I've updated the test to compare the arrays

File: numba/np/arrayobj.py [hunk 10]
```diff
@@ -5174,6 +5174,32 @@ def array_to_array(context, builder, fromty, toty, val):
...
+
+    sig = signature(toty, fromty)
+    res = context.compile_internal(builder, impl, sig, [val])
+    return impl_ret_untracked(context, builder, sig.return_type, res)
+
+
+@lower_cast(types.Array, types.UnicodeCharSeq)
+def array_to_unichrseq(context, builder, fromty, toty, val):
+    def impl(a):
+        return str(a[()])
```
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply numba/np/arrayobj.py:5196]: Yes! For the following test to work
```python
@njit
foo(idx, a, b):
    a[idx] = b

foo(0, np.zeros((3,), dtype='<U3'), np.array('a'))

```

**Comment by guilhermeleobas** [comment]
@stuartarchibald, I've included a test for `Array[Unicode] -> Unicode`

File: numba/tests/test_fancy_indexing.py [hunk 11]
```diff
@@ -181,47 +181,22 @@ def test_setitem_0d(self):
...
-
+        inps = [
+            (np.zeros(3), np.array(3.14)),
+            (np.zeros(2), np.array(2)),
+            (np.zeros(3, dtype=np.int64), np.array(3, dtype=np.int64)),
+            (np.zeros(3, dtype=np.float64), np.array(1, dtype=np.int64)),  # coercing
+            (np.zeros(5, dtype='<U3'), np.array('abc')),  # UnicodeCharSeq
+            (np.zeros((3,), dtype='<U3'), np.array('a')),
+            (np.array(['abc','def','ghi'], dtype='<U3'), np.array('WXYZ', dtype='<U4')),
+            (np.zeros(3, dtype=complex), np.array(2+3j, dtype=complex)),  # complex
```
Comment (stuartarchibald) [inline numba/tests/test_fancy_indexing.py:192]:
```suggestion
            (np.zeros(3, dtype=np.int64), np.array(3, dtype=np.int64)),
            (np.zeros(3, dtype=np.float64), np.array(1, dtype=np.int64)),
            (np.zeros(5, dtype='<U3'), np.array('abc')),
            (np.zeros((3,), dtype='<U3'), np.array('a')),
            (np.array(['abc','def','ghi'], dtype='<U3'),
             np.array('WXYZ', dtype='<U4')),
            (np.zeros(3, dtype=complex), np.array(2+3j, dtype=complex)),
```
long lines

**Comment by stuartarchibald** [comment]
> @stuartarchibald, I've included a test for `Array[Unicode] -> Unicode`

Thanks @guilhermeleobas, I've take a look at 6c01ccc and I think it's ok now (apart from the long lines as noted above). I'm still thinking about whether it's always valid to convert unicode char sequence to the same irrespective of `count`, but I think for now what is present is reasonable until proven otherwise :)

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch!
